pycbc/inference/burn_in.py

```diff
@@ -28,6 +28,59 @@
 
 import numpy
 
+def ks_test(sampler, fp):
+    """Burn in based on whether the p-value of the KS test between the samples
+    at the last iteration and the samples midway along the chain for each
+    parameter is > 0.1 and < 0.9
+
+    Parameters
+    ----------
+    sampler : pycbc.inference.sampler
+        Sampler to determine burn in for. May be either an instance of a
+        `inference.sampler`, or the class itself.
+    fp : InferenceFile
+        Open inference hdf file containing the samples to load for determing
+        burn in.
+
+    Returns
+    -------
+    burn_in_idx : array
+        Array of indices giving the burn-in index for each chain.
+    is_burned_in : array
+        Array of booleans indicating whether each chain is burned in.
+    """
+    from scipy.stats import ks_2samp
+    from scipy import interpolate
+    nwalkers = fp.nwalkers
+    niterations = fp.niterations
+    burn_in_idx = numpy.repeat((niterations-1), nwalkers).astype(int)
+    # The chains should not be burned in by default. So is_burned_in is assigned
+    # a False value for all the walkers by default.
+    is_burned_in = numpy.zeros(nwalkers, dtype=bool)
+    # Create a dictionary which would have keys are the variable args and values
+    # are booleans indicating whether the p-value for the parameters satisfies
+    # the KS test
+    is_burned_in_param = {}
+    # iterate over the parameters
+    for param in fp.variable_args:
+        # read samples for the parameter from the last iteration of the chain
+        samples_last_iter = sampler.read_samples(fp, param, iteration=-1,
+                                                 flatten=True)[param]
+        # read samples for the parameter from the iteration midway along the chain
+        samples_chain_midpt = sampler.read_samples(fp, param, iteration=int(niterations/2),
+                                                  flatten=True)[param]
+        _, p_value = ks_2samp(samples_last_iter, samples_chain_midpt)
+        # check if p_value is within the desired range
+        if p_value > 0.1 and p_value < 0.9 :
+            is_burned_in_param[param] = True
+        else :
+            is_burned_in_param[param] = False
+    # The chains are burned in if the p-value of the KS test lies in the range [0.1,0.9]
+    # for all the parameters
+    if numpy.all([is_burned_in_param[x] for x in is_burned_in_param]) :
```

Comment:
Set both `is_burned_in` and `burn_in_idx` here. If it is burned in, set the burn in idx to be the mid-way point (since the ks test shows that they are the same, then the sampler must have been burned in by the mid-way point), whereas if it is not burned in, set it to the number of iterations. Also, you can just use python's native `all` instead of numpy's:
```
if all(is_burned_in_param.values()):
    is_burned_in = numpy.ones(nwalkers, dtype=bool)
    burn_in_idx = numpy.repeat(niterations/2, nwalkers).astype(int)
else:
    is_burned_in = numpy.zeros(nwalkers, dtype=bool)
    burn_in_idx = numpy.repeat(niterations, nwalkers).astype(int)
```

Reply:
That makes sense.